plasmapy/diagnostics/langmuir.py

```diff
@@ -1,153 +1,1073 @@
-"""
-Defines the Langmuir analysis module as part of the diagnostics package.
-"""
+"""Defines the Langmuir analysis module as part of the diagnostics package."""
 
-import numpy as np
+import plasmapy.constants as const
 import astropy.units as u
+import copy
+
+from plasmapy import utils
+
+from astropy.visualization import quantity_support
+import numpy as np
+import matplotlib.pyplot as plt
+
 
+def fit_func_lin(x, x0, y0, c0):
+    r"""Linear fitting function."""
+    return y0 + c0 * (x - x0)
 
-class Probe():
-    r"""
-    Class containing various Langmuir probe properties.
 
-    For the moment cylindrical probes are assumed as spherical ones are
-    generally impractical to make.
+def fit_func_double_lin(x, x0, y0, c0, c1):
+    r"""Piecewise linear fitting function. (x0, y0) denotes the location of
+    the knee of the transition, with c0 and c1 being the left and right-hand
+    slopes, respectively.
+    """
+    return np.piecewise(x, x < x0, [lambda x: y0 + c0 * (x - x0),
+                                    lambda x: y0 + c1 * (x - x0)])
+
+
+class Characteristic:
+    r"""Class representing a I-V probe characteristic for convenient access.
+    Supports units.
 
     Attributes
     ----------
-    area : ~astropy.units.Quantity or ~numpy.ndarray
-        Area or array of areas of the probe(s) in units convertible to m^2.
+    bias : Quantity, ndarray
+        Array of applied probe biases in units convertible to V.
 
-    config : str, optional
-        Electrode configuration. Possible options are `'single'`
-        (default), `'double'`, `'double-asym'`, `'triple'`, `'tetra'`,
-        `'penta'`. The electrode configuration must correspond to the
-        number of areas provided.
+    current : Quantity, ndarray
+        Array of applied probe currents in units convertible to A.
 
     """
-    def __init__(self, area=u.m**2, config='single'):
-        # Initialize probe properties
-        self.area = area
-        self.config = config
 
+    def __init__(self, bias, current):
+        self.bias = bias
+        self.current = current
+
+        self.check_validity()
+
+    def __getitem__(self, key):
+        r"""Allow array indexing operations directly on the Characteristic
+        object.
+
+        """
+
+        b = Characteristic(self.bias[key], self.current[key])
+        return b
+
+    def __sub__(self, other):
+        r"""Support characteristic subtraction"""
+
+        b = copy.deepcopy(self)
+        b.current -= other.current
+        return b
+
+    def __add__(self, other):
+        r"""Support characteristic addition"""
+
+        b = copy.deepcopy(self)
+        b.current += other.current
+        return b
+
+    def sort(self):
+        r"""Sort the characteristic into an ascending bias."""
+
+        _sort = self.bias.argsort()
+        self.current = self.current[_sort]
+        self.bias = self.bias[_sort]
+
+    def get_unique(self):
```

Comment:
Would it be useful to add an `inplace` argument, that instead of returning a new `Characteristic`, modifies the existing one?